Add Diagram Guide to Documentation Style Overview section (#30299)
kubernetes/website: content/en/docs/contribute/style/diagram-guide.md

---
title: Diagram Guide
linktitle: Diagram guide
content_type: concept
weight: 15
---

<!--Overview-->

This guide shows you how to create, edit and share diagrams using the Mermaid javascript library. Mermaid.js allows you to generate diagrams using a simple markdown-like syntax inside markdown files. You can also use Mermaid to generate `.svg` or `.png` image files that you can add to your documentation.

The target audience for this guide is anybody wishing to learn about Mermaid and/or how to create and add figures to Kubernetes documentation.

The following diagram outlines the topics covered in this section. 

{{< mermaid >}}
flowchart LR
subgraph m[Mermaid.js]
direction TB
S[ ]-.-
C[build<br>diagrams<br>with markdown] -->
D[on-line<br>live editor]
end
A[Why are diagrams<br>useful?] --> m
m --> N[3 x methods<br>for creating<br>diagrams]
N --> T[Diagram<br>styling]
T --> X[Examples and<br>tips]

    classDef box fill:#fff,stroke:#000,stroke-width:1px,color:#000;
    classDef spacewhite fill:#ffffff,stroke:#fff,stroke-width:0px,color:#000
    class A,C,D,N,X,m,T box
    class S spacewhite
{{< /mermaid >}}

> <figure alt="diagram guide topics covered" style="padding-left: 310px; font-weight: bold">

Reviewer: Is there a way we can use a CSS class rather than inline styles?
(could be a follow-up PR)

Author: Prefer follow-up PR.

The user benefits include:

* __Friendly landing spot__. A detailed text-only greeting page could intimidate users, in particular, first-time Kubernetes users. 
* __Faster grasp of concepts__. A diagram can help users understand the key points of a complex topic. Your diagram can serve as a visual learning guide to dive into the topic details. 
* __Better retention__. For some, it is easier to recall pictures rather than text.

The contributor benefits include:

* __Assist in developing the structure and content__ of your contribution. For example, you can start with a simple diagram covering the high-level points and then dive into details.  
* __Expand and grow the user community__. Easily consumed documentation augmented with diagrams attracts new users who might previously been reluctant due to perceived complexities. 

You should consider your target audience. In addition to experienced K8s users, you will have many who are new to Kubernetes. Even a simple diagram can assist new users in absorbing Kubernetes topics. They become emboldened and more confident to further explore Kubernetes and the documentation.  

## Mermaid

> [Mermaid](https://mermaid-js.github.io/mermaid/#/) is an open source javascript library that allows you to create, edit and easily share diagrams using a simple, markdown-like syntax configured inline in markdown files. 

Reviewer: (only nits)
```suggestion
[Mermaid](https://mermaid-js.github.io/mermaid/#/) is an open source JavaScript library that allows you to create, edit and easily share diagrams using a simple, markdown-like syntax configured inline in Markdown files.
```

> The following lists features of Mermaid. 

Reviewer: nit:
```suggestion
The following lists features of Mermaid:
```

* Simple code syntax.

> * Already supported and used in [K8s.io](https://kubernetes.io/). 

Reviewer: ```suggestion
Maybe omit this line, given where we're writing?
```

* Includes a web-based tool allowing you to code and preview your diagrams.
* Supports multiple formats including flowchart, state and sequence.
* Easy collaboration with colleagues by sharing per-diagram URL. 
* Broad selection of shapes, lines, themes and styling. 

> The following lists advantages of using Mermaid.

Reviewer: nit:
```suggestion
The following lists advantages of using Mermaid:
```

</figure>

For example, use the live editor to create a diagram called `boxnet`. Store the diagram URL somewhere for later reference. Generate and download a `boxnet.svg` file to the appropriate `../images/` folder. Then reference the `.svg` file in the PR `.md` file using the standard markdown image syntax.

```json
![my boxnet diagram](static/images/boxnet.svg)
```

You can also use the [Hugo `{{</* figure */>}}` shortcode](https://gohugo.io/content-management/shortcodes/#figure) to reference a diagram. It allows you to include a title, caption, various link URLs and styling associated with your diagram. 

```
{{</* figure src="/static/images/boxnet.svg" title="Boxnet figure" caption="Picture of Boxnet" */>}}
```

> 

Reviewer: ```suggestion

Once you have created this SVG image, best practice is to load it into an SVG editing tool and use the
“Convert text to paths” function. Making that change ensures that the image renders the same on all systems,
regardless of font availability and font rendering support.
Save the updated image and use the revised version in the next steps.

```

```
<!-- To view or edit the mermaid code, use the following URL: -->
<!-- https://mermaid-js.github.io/mermaid-live-editor/edit/#eyJjb ... -->
```

The advantages of the Mermaid+SVG method consist of the following:

* Live editor tool.
* Live editor tool supports the most current Mermaid feature set.  
* Employ existing K8s/website methods for handling `.svg` file.
* Environment doesn't require Mermaid support.

Be sure to check the diagram renders properly using the local and Netlify previews.

**3. External Tool** - Use your external tool to create a diagram. Save the diagram coordinates for contributor access. For example, your tool may offer a link to the diagram image, or you could place the source code file, such as an `.xml` file, in a public repository for contributor access.   

Generate an `.svg` or `.png` file, download this file to the appropriate `../images/` folder, and then reference the `.svg` or `.png` file using a markdown image statement.

Here is an example markdown image statement for `your-image.svg`. 

```json
![your-image.svg](static/images/boxnet.svg)

> ```

Reviewer: If it's a diagram, I'd prefer that people use a [`figure`](https://gohugo.io/content-management/shortcodes/#figure) shortcode and that they additionally set a size class.
For example:
```markdown
{{< figure src="/images/docs/pod.svg" alt="Pod creation diagram" class="diagram-medium" >}}
```

Allowed sizes are `small`, `medium` or `large`.

    classDef box fill:#fff,stroke:#000,stroke-width:1px,color:#000;
    class A,B,C,D,E box
{{< /mermaid >}}

<figure style="padding-left: 310px; font-weight: bold" alt="diagram guide external tool method"> 
<figcaption><a href="https://mermaid-js.github.io/mermaid-live-editor/edit/#eyJjb2RlIjoiZ3JhcGggTFJcbiAgICBBWzEuIFVzZSBleHRlcm5hbDxicj50b29sIHRvIGNyZWF0ZS9lZGl0PGJyPmRpYWdyYW1dIC0tPlxuICAgIEJbMi4gSWYgcG9zc2libGUsIHNhdmU8YnI-ZGlhZ3JhbSBjb29yZGluYXRlczxicj5mb3IgY29udHJpYnV0b3I8YnI-YWNjZXNzXSAtLT5cbiAgICBDWzMuIEdlbmVyYXRlIC5zdmcgPGJyPm9yLnBuZyBmaWxlPGJyPmFuZCBkb3dubG9hZCB0bzxicj5hcHByb3ByaWF0ZTxicj5pbWFnZXMvIGZvbGRlcl0gLS0-XG4gICAgRFs0LiBSZWZlcmVuY2UgLnN2ZyBvciAucG5nIGZpbGU8YnI-dXNpbmcgbWFya2Rvd24gaW1hZ2U8YnI-c3RhdGVtZW50IGluIHBhZ2UgLm1kIGZpbGUgcGFnZV1cblxuXG4gICAgY2xhc3NEZWYgYm94IGZpbGw6I2ZmZixzdHJva2U6IzAwMCxzdHJva2Utd2lkdGg6MXB4LGNvbG9yOiMwMDA7XG4gICAgY2xhc3MgQSxCLEMsRCxFIGJveFxuICAgICIsIm1lcm1haWQiOiJ7XG4gIFwidGhlbWVcIjogXCJkZWZhdWx0XCJcbn0iLCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9">Figure x. External Tool method steps</a></figcaption><br>
</figure>

If your external drawing tool permits:

> * You can incorporate multiple `.svg` or `.png` logos, icons and images into your diagram. However, make sure you follow K8s/website guidelines on the use of [third party content](https://kubernetes.io/docs/contribute/style/content-guide/).

Reviewer: ```suggestion
* You can incorporate multiple `.svg` or `.png` logos, icons and images into your diagram.
  However, make sure you observe copyright and follow the Kubernetes documentation
  [guidelines](/docs/contribute/style/content-guide/) on the use of third party content.
```

* You should save the diagram coordinates for contributor access. For example, your tool may offer a link to the diagram image, or you could place the source code file, such as an `.xml` file, somewhere for contributor access.

For more information on K8s and CNCF logos and images, check out [CNCF Artwork](https://github.com/cncf/artwork).

The advantages of the external tool method consist of the following: 

* Contributor familiarity with external tool.
* Diagrams require more detail than what Mermaid can offer.

> Don't forget to check the rendered diagram using the local and Netlify previews.

Reviewer: nit: maybe also link to https://kubernetes.io/docs/contribute/new-content/open-a-pr/#preview-locally

Without captions, you are asking the user to scan the text above or below the diagram to figure out a meaning. This can be frustrating for the user. 

{{< note >}}
You should always add a caption to each diagram in your documentation.
{{< /note >}}

Proper captioning involves three components: diagram, diagram caption and the diagram referral. As an example, consider a diagram `.svg` image file saved to `/images/docs/components-of-kubernetes.svg`.

**Diagram**

Here is the `{{</* figure */>}}` shortcode for the diagram:
```text
{{</* figure src="/images/docs/components-of-kubernetes.svg" alt="Kubernetes pod running inside a cluster" class="diagram-large" */>}}
```
You should only pass the `src`, `alt` and `class` values into the `{{</* figure */>}}` shortcode. You can adjust the size of the diagram using `diagram-large`, `diagram-medium` and `diagram-small` classes. You can't style the `caption` value in this shortcode, and it only supports diagram image files. 

{{< note >}}
Diagrams created using the `Inline` method don't use the `{{</* figure */>}}` shortcode.
{{< /note >}}

**Diagram Caption**

> Next, add a diagram caption. You must use the following format:

Reviewer: I would prefer to recommend use the caption support built into Hugo:
```markdown
{{< figure src="/images/docs/components-of-kubernetes.svg"
 alt="One Kubernetes pod running inside a cluster" class="diagram-large"
 caption="The new pod running inside your cluster"
>}}
```

```text
<figcaption><p style="padding-left: 310px; font-weight: bold">Figure #. CAPTION TEXT</p></figcaption><br>
```
Here are several items to consider when adding diagram captions: 

* Use`<figcation></figcaption>` tags to add a diagram caption for `Inline`, `Mermaid+SVG` and `External Tool` diagrams. 
* Wrap caption text inside `<p></p>` tags using `style="padding-left: 310px; font-weight: bold"`. This provides a consistent style for the caption text.

> * Prepend your diagram caption with `Figure NUMBER.`. You must use `Figure` and the number must be unique for each diagram in your documentation page. 

Reviewer: This is the sort of thing that Hugo can compute for us, if we want it to. Let's not make authors do the work here.

* Add your diagram caption text after the `Figure NUMBER.` on the same line. Keep it short.
* Position your diagram caption __BELOW__ your diagram.
* You can hyperlink your caption using `<a></a>` tags.

Here is the caption for the `/images/docs/components-of-kubernetes.svg` diagram: 
```text
<figcaption><p style="padding-left: 310px; font-weight: bold">Figure 4. Kubernetes Architecture.</p></figcaption><br>
```
Here is the hyperlinked caption for the same diagram:
```text
<figcaption><a href="https://kubernetes.io/docs/concepts/overview/components/"><p style="padding-left: 310px; font-weight: bold">Figure 4. Kubernetes Architecture Test.</p></a></figcaption><br>
```

Check out the source code for this page, `diagram-guide.md`, for more examples of diagram captions.

**Diagram Referral**

Finally, you can add a diagram caption referral. This is used inside your text. It allows a user to connect your text with the diagram. The `Figure NUMBER` in your referral and caption must match.   

Here is an example of a caption referral included in your content text:
```text
Figure 9 depicts the components of the Kubernetes architecture. The control plane ...
```

> Diagram referrals are optional and there are cases where they might not be suitable. If you are not sure, add a diagram referral to your text to see if it looks and sounds okay. When in doubt, use a diagram referral.

Reviewer: We can use hypertext to link to diagrams, rather than numbered figures. Let's omit this and if we find we need these, add a hypertext mechanism instead.

**Complete picture**

Figure 9 shows the Kubernetes Architecture diagram that includes the diagram, diagram caption and diagram caption referral. The `{{</* figure */>}}` shortcode renders the diagram. The `<figcaption>` defines the caption. The caption referral is contained in this paragraph.

Here are the `{{</* figure */>}}` shortcode and `<figcaption>` statements for this diagram:
```
{{</* figure src="/images/docs/components-of-kubernetes.svg" alt="Kubernetes pod running inside a cluster" class="diagram-large" */>}}
<figcaption><a href="https://kubernetes.io/docs/concepts/overview/components/"><p style="padding-left: 310px; font-weight: bold">Figure x. Kubernetes Architecture Test.</p></a></figcaption><br>
```

{{< figure src="/images/docs/components-of-kubernetes.svg" alt="Kubernetes pod running inside a cluster" class="diagram-large" >}}

> <figcaption><p style="padding-left: 310px; font-weight: bold">Figure 9. Kubernetes Architecture.</p></figcaption><br>

Reviewer: It's not good practice to have a `<figcaption>` outside of the `<figure>` element it's for. See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/figcaption

Author: Yes, aware of that. My thinking was:

- We don't have a consistent/suitable caption format for all diagram types. We need a shortcode for this.
- we want to avoid any more html than what is necessary, like <figure> tags. Thinking as someone who has never seen html/css. 
- <figcaption> works as "placeholder" for new "{{< caption >}} shortcode."

Suggest we keep as is, then open PR for new caption shortcode? Or revise text so that
```
<figure><figcaption> ... </figcaption></figure>
``` 
is used now and new caption shortcode PR follows. Guidance needed.

Reviewer: Let's save that _discussion_ for the new issue / PR. I definitely recommend documenting `{{< figure >}}` shortcodes, with no extra HTML, for now.

BTW https://gohugo.io/content-management/shortcodes/#figure does already support a `caption` attribute.

Author: `{{< figure >}}` shortcode is in PR there for `.svg` cases. However it does not apply to the `Inline` method since there is no `.svg` image file.  This shortcode's caption attribute left justify's the caption text. This does not look right. 

Figure captions (Figure Number. Caption text) are essential for good technical documentation. IMHO, we need a uniform style and method right now for contributors to include figure captions for the `Inline`, `Mermaid+SVG` and `External Tool` methods. 

The only solution that I see right now is to tell contributors to cut/paste this code to insert captions:
```
<figure><figcaption><p style="padding-left: 310px; font-weight: bold">Figure #. CAPTION TEXT</p></figcaption></figure><br>
```
and of course update the Figure #. CAPTION TEXT portion.

Reviewer: I firmly recommend saving the discussion on captioning for a follow up PR. That way, we can land the rest of the changes and then look into the best way to set up effective captioning, including how we align and style the text of the caption.

Reviewer: To clear on what I want to _omit_: advice to readers about adding an invalid, standalone `<figcaption>`. If _this_ page uses standalone `<figcaption>` elements that's not great; we can accept it and plan to revisit.

But if we are telling everyone who reads the page to do the wrong thing, that takes much more effort to rework. I don't want to merge a page that gives advice that that I would immediately want to strike out.

Author: New contributor, diagram advocate hat on: :-)
- diagram captions are a must. Sporadically used through out docs. Figure shortcode is fine for `.svg` files. *We have nothing for Mermaid diagrams.*
- diagram referrals (e.g. Figure 9 illustrates ...) needed. Hard link between some text and actual figure. [Spatial](https://developers.google.com/style/images#figure-captions) approaches are buried in the text somewhere forcing users to jump around. Lots of text in the docs omit or use spatial approaches.
- I understand best practice use of <figure><figcaption> ... tags.
- We need single, uniform, consistent, documented method for adding figure captions.
New contributor, diagram advocate hat off: :-)

To progress this much needed PR: 
- Recommend `.svg` figure captions use figure shortcode caption attribute.
- Recommend `Inline` method Mermaid diagrams use text.

We can later on open PR to fix figure shortcode to handle Mermaid diagrams. How does that sound?

Reviewer: SGTM @chrismetz09

Author: Okay got it! Tx for the lengthy discussion @sftim. Commits on the way.
